notebook/converted/visualize_errors.py

```diff
@@ -0,0 +1,848 @@
+
+
+# ### Sequencer Results
+# 
+# There seems to be a marked increase in hard-to-classify Hi-seq 2000 samples and decrease in Hi-seq 2500 samples. Is there a reason for this?
+# 
+
+# In[158]:
+
+
+hard_studies = set(hard_df['study'])
+hard_study_df = metadata_df[metadata_df['study'].isin(hard_studies)]
+
+
+# In[159]:
+
+
+hard_study_df['sra.platform_model'].value_counts() 
+
+
+# In[160]:
+
+
+hard_study_df['sra.platform_model'].value_counts() / len(hard_study_df)
+
+
+# ### More sequencer results
+# The enrichment for Hi-seq 2000 may just be because the studies with hard samples are themselves enriched for Hi-seq 2000. 
+# The sequencer probably may not be meaningful, but this is still interesting. It feels like the samples are drawn at random from the studies, yielding the given distribution.
+# 
+# Unclear whether this is true though, the studies themselves seem to to be heavily skewed.
+
+# ## Remaining questions:
+# - Is it related to sparsity?
+# - Is there something correlated with sequencer that we'd expect to cause these results?
+# - Stuff below
+
+# ## Error analysis on hard samples
+
+# In[161]:
+
+
+sample_model_predictions = {}
+model_predictions = {}
+# Process dataframe rows
+for i, row in tissue_metrics[tissue_metrics['is_pretrained'] == 'not_pretrained'].iterrows():
+    model = row['supervised']
+    
+    if model not in model_predictions:
+        model_predictions[model] = {'predicted': [], 'true': [], 'sample': []}
+    
+    encoder_string = row['val_encoders']
+    encoder = json.loads(encoder_string)
+    decoder = {number: label for label, number in encoder.items()}
+    
+    samples = row['val samples'].strip().split(',')
+    
+    predictions = row['val_predictions'].strip().split(',')
+    truth = row['val_true_labels'].strip().split(',')
+    pred_labels = []
+    true_labels = []
+
+    for prediction, true_label, sample in zip(predictions, truth, samples):
+        
+        if sample not in hard_samples:
+            continue
+        
+        if int(prediction) in decoder:
+            pred_labels.append(decoder[int(prediction)])
+        else:
+            # https://github.com/greenelab/saged/issues/58
+            pred_labels.append('invalid_index')
+            
+        true_labels.append(decoder[int(true_label)])
+               
+    predictions = pred_labels
+    truth = true_labels
+        
+    assert len(truth) == len(predictions)
+    
+    model_predictions[model]['predicted'].extend(predictions)
+    model_predictions[model]['true'].extend(truth)
+    model_predictions[model]['sample'].extend(samples)
+
+
+# In[162]:
+
+
+confusion_df = create_confusion_df(model_predictions)
+
+
+# In[163]:
+
+
+melted_df = confusion_df.melt(id_vars='true_tissue', var_name='pred_tissue')
+melted_df['log_value'] = np.log(melted_df['value'])
+melted_df['percent'] = (melted_df['value'] * 100).round(1)
+
+
+# In[164]:
+
+
+plot = ggplot(melted_df, aes(x='true_tissue', y='pred_tissue', fill='percent',)) 
```

Comment:
Have you thought about whether this confusion matrix makes biological sense? For example, maybe it makes sense that some blood cells are transcriptionally closer to stem cells, or that some brain cells look like skin cells (I don't know enough about development/cell lineages to say whether these are true, just ideas that I see in your plot).

Reply:
I would expect that to be part of the signal, but the outliers seem unrelated. For example, everything gets predicted to be lung tissue or spinal cord.  While I'm not sure what causes those horizontal streaks, some of the errors I think are biological, like the bone marrow <-> blood mixups